pkg/controllers/user/globaldns/globaldns_common.go

```diff
@@ -0,0 +1,102 @@
+package globaldns
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"k8s.io/api/core/v1"
+)
+
+func gatherIngressEndpoints(ingressEps []v1.LoadBalancerIngress) []string {
+	endpoints := []string{}
+	for _, ep := range ingressEps {
+		if ep.IP != "" {
+			endpoints = append(endpoints, ep.IP)
+		} else if ep.Hostname != "" {
+			endpoints = append(endpoints, ep.Hostname)
+		}
+	}
+	return endpoints
+}
+
+func getMultiClusterAppName(multiClusterAppName string) (string, error) {
+	split := strings.SplitN(multiClusterAppName, ":", 2)
+	if len(split) != 2 {
+		return "", fmt.Errorf("error in splitting multiclusterapp ID %v", multiClusterAppName)
+	}
+	mcappName := split[1]
+	return mcappName, nil
+}
+
+func prepareGlobalDNSForUpdate(globalDNS *v3.GlobalDNS, ingressEndpoints []string, clusterName string) {
+	originalLen := len(globalDNS.Status.Endpoints)
+	globalDNS.Status.Endpoints = append(globalDNS.Status.Endpoints, ingressEndpoints...)
+
+	if originalLen > 0 {
+		//dedup the endpoints
+		globalDNS.Status.Endpoints = dedupEndpoints(globalDNS.Status.Endpoints)
+	}
+
+	//update clusterEndpoints for current cluster
+	if len(globalDNS.Status.ClusterEndpoints) == 0 {
+		globalDNS.Status.ClusterEndpoints = make(map[string][]string)
+	}
+
+	clusterEps := globalDNS.Status.ClusterEndpoints[clusterName]
+
+	if ifEndpointsDiffer(clusterEps, ingressEndpoints) {
```

Comment:
do we only call this method - `ifEndpointsDiffer` - for the cluster endpoints update? should be called on the aggregated endpoints update as well.

Reply:
done